Add support for Event Bridge

jameshopkins opened the pull request:
Adds support for Event Bridge PutEvents.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/k6-jslib-aws?pullRequest=55) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/k6-jslib-aws?pullRequest=55) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/k6-jslib-aws?pullRequest=55) it.</sub>

oleiade commented on src/internal/event-bridge.ts:
```diff
@@ -0,0 +1,139 @@
+
+type PutEventsInput = {
+    EndpointId?: string
+    Entries: PutEventEntry[]
```
Based on the fact that the [`PutEventsRequestEntry` type](https://docs.aws.amazon.com/eventbridge/latest/APIReference/API_PutEventsRequestEntry.html) defines all its fields as optional, let's use the following definition instead:

```suggestion
    Entries: Partial<PutEventEntry>[]
```

So that it's possible to effectively pass in a `PutEventEntry` with partially undefined properties.

jameshopkins replied:
That link does however state that `Source`, `Detail` and `DetailType` are required fields to send events to Event Bridge, but the others aren't. So if you're in agreement, I'll set these as required, but the others as optional.

oleiade replied:
This is my bad. That will teach me to not read the excerpt with the blue background 🤓 My understanding was based on the `Required: no` statements under each field. 

That makes sense, then, and let's go forward with setting only the non-necessary fields as optional 🙇🏻

oleiade commented on src/internal/event-bridge.ts:
```diff
@@ -0,0 +1,139 @@
+import { HTTPHeaders, HTTPMethod } from './http'
+
+type PutEventEntry = {
+    Detail: string
```
The [documentation](https://docs.aws.amazon.com/eventbridge/latest/APIReference/API_PutEventsRequestEntry.html) states "A valid JSON object. There is no other schema imposed. The JSON object may contain fields and nested sub-objects.".

Later in the code, you `JSON.stringify` this field lets me think you expect an actual `Object` to be passed in. 

Judging by how this will be used (disclaimer: I haven't used event bridge yet), I believe the fact that it is sent as a string to the AWS API is an implementation detail and that exposing it to the user as an `Object` might lead to a better developer experience.

I'm not certain we should change it, but for reference, I believe we could take this opportunity to harness the typescript type system power and define a type for it:

```typescript
type JsonValue = string | number | boolean | null | JsonArray | JsonObject;
type JsonArray = JsonValue[];
interface JsonObject { [key: string]: JsonValue; }

interface PutEventEntry {
    Detail: JsonObject;  // or JsonValue if you want to be more permissive
    DetailType: string;
    EventBusName: string;
    Resources: string[];
    Source: string;
}
```

This introduces some abstraction for an allegedly small benefit, and once again I'm not a user, so I don't know if it would feel better that way.

What do you think?

jameshopkins replied:
Good spot! It was a design decision to have `Detail` as a JSON object type (see the test), but I had specified the wrong type of `string`. I'll update as per your suggestion here 👍

oleiade replied:
Great, thanks you 🙏🏻

oleiade commented on src/internal/event-bridge.ts:
```diff
@@ -0,0 +1,139 @@
+    Source: string
+}
+
+type PutEventsInput = {
```
I'm not opposed to using type definitions per-se. However, it's been our practice so far to use interface types instead when defining a "contract" object that's meant to represent function arguments.

I would suggest the following (including documentation):

```suggestion
/**
 * Represents the input for a put events operation.
 *
 * @typedef {Object} PutEventsInput
 *
 * @property {string} [EndpointId] - The optional URL subdomain of the endpoint.
 * @property {Partial<PutEventEntry>[]} Entries - An array of entries that defines an event in your system.
 */
interface PutEventsInput {
    EndpointId?: string
    Entries: Partial<PutEventEntry>[]
}
```

What do you think?

oleiade commented on src/internal/event-bridge.ts:
```diff
@@ -0,0 +1,139 @@
+import { AMZ_TARGET_HEADER } from './constants'
+import { HTTPHeaders, HTTPMethod } from './http'
+
+type PutEventEntry = {
```
Same comment regarding using interface type.

Here's my suggestion with documentation:
```suggestion
/**
 * Represents an event to be submitted.
 *
 * @typedef {Object} PutEventEntry
 *
 * @property {string} Detail - A valid serialized JSON object. There is no other schema imposed. The JSON object may contain fields and nested sub-objects.
 * @property {string} DetailType - Free-form string, with a maximum of 128 characters, used to decide what fields to expect in the event detail.
 * @property {string} EventBusName - The name or ARN of the event bus to receive the event. Only the rules that are associated with this event bus are used to match the event. If you omit this, the default event bus is used.
 * @property {string[]} Resources - AWS resources, identified by Amazon Resource Name (ARN), which the event primarily concerns. Any number, including zero, may be present.
 * @property {string} Source - The source of the event.
 */
interface PutEventEntry {
    Detail: string
    DetailType: string
    EventBusName: string
    Resources: [string]
    Source: string
}
```

What do you think?

oleiade reviewed (commented):
Thanks a lot for your contribution, much appreciated 🎉 🙇🏻 Great work overall 👏🏻 

I've left a couple of change requests that are just to align the code with our common practices and document it to make my life easier later on as I amend the k6 documentation.

oleiade commented:
Hey @jameshopkins 👋🏻 

Do you happen to have any updates on this? If you don't have the time or capacity, that's perfectly fine. please just let me know, and I will probably be able to address my comments myself 🙇🏻

jameshopkins commented:
Hey @oleiade  really sorry I haven’t got back re progress on this before now. Work has been pretty full-on.

would you mind addressing the comments if you’re able to? Sorry I wouldn’t normally do this but I have very little time right now

oleiade commented:
Sure thing, will address them indeed 👍🏻

oleiade commented:
I have modified the PR and merged it indeed. Thanks a lot for your contribution again 🎉

oleiade commented:
I have created a [PR](https://github.com/grafana/k6-docs/pull/1348) for this in the documentation website's repository.
